exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+	infraMachineList.SetKind(infraMachineKind)
+
+	sampleInfraMachine := &unstructured.Unstructured{}
+	sampleInfraMachine.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	sampleInfraMachine.SetKind(infraMachineKind)
+	// Add watcher for infraMachine, if there isn't one already.
+	// Ensure we add a watch to the infraMachine, if there isn't one already.
+	if err := r.externalTracker.Watch(log, sampleInfraMachine, handler.EnqueueRequestsFromMapFunc(infraMachineToMachinePoolMapper)); err != nil {
+		return errors.Wrapf(err, "failed to add watcher on infraMachine %q", sampleInfraMachine.GroupVersionKind())
+	}
+
+	if err := r.Client.List(ctx, &infraMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(infraMachineSelector.MatchLabels)); err != nil {
+		return errors.Wrapf(err, "failed to list infra machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		clusterv1.MachinePoolNameLabel: mp.Name,
+		clusterv1.ClusterNameLabel:     mp.Spec.ClusterName,
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	if err := r.deleteMachinesWithoutInfrastructure(ctx, mp, machineList.Items, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to clean up orphan machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	updatedMachines, err := r.createMachinesIfNotExists(ctx, mp, machineList.Items, infraMachineList.Items)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.ensureInfraMachineOnwerRefs(ctx, updatedMachines, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	return nil
+}
+
+// deleteMachinesWithoutInfrastructure deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteMachinesWithoutInfrastructure(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting machines without infrastructure", "machinePool", mp.Name, "namespace", mp.Namespace)
+
+	infraMachineNames := sets.Set[string]{}
+	for _, infraMachine := range infraMachines {
+		infraMachineNames.Insert(infraMachine.GetName())
+	}
+
+	for i := range machines {
+		machine := &machines[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will fail if the infraRef is empty, i.e. no object kind or apiVersion.
+		// Delete the machine if the infraRef is empty.
+		if (infraRef == corev1.ObjectReference{}) {
```

Comment:
How is this case possible?

Machine defaulting webhook
```go

	if m.Spec.InfrastructureRef.Namespace == "" {
		m.Spec.InfrastructureRef.Namespace = m.Namespace
	}
```

Reply:
Originally, this was an expected case when we had fallback Machines, i.e. Machines w/o any infrastructure for InfraMachinePools that don't support MPMs. But in this case, I think we just want to be explicit because otherwise the MachinePool would be unable to recover.

Reply:
But looking at the current code, this is unreachable code. I would really prefer if we not handle cases that are unreachable.

Reply:
Fair enough. I'll replace it with a note saying that we need this check if we decide to support MPMs without infrastructure.